File: android/app/src/main/java/com/alexivanov/snapsell/ui/handoff/HandOffScreen.kt
package com.alexivanov.snapsell.ui.handoff

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.compose.LifecycleEventEffect
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.R
import com.alexivanov.snapsell.ui.common.ErrorText
import com.alexivanov.snapsell.ui.common.GhostButton
import com.alexivanov.snapsell.ui.common.LoadingBox
import com.alexivanov.snapsell.ui.common.LucideIcon
import com.alexivanov.snapsell.ui.common.MicroLabel
import com.alexivanov.snapsell.ui.common.PrimaryButton
import com.alexivanov.snapsell.ui.common.Rule
import com.alexivanov.snapsell.ui.common.Scrim
import com.alexivanov.snapsell.ui.common.SecondaryButton
import com.alexivanov.snapsell.ui.common.SnapBottomBar
import com.alexivanov.snapsell.ui.common.SnapTopBar
import com.alexivanov.snapsell.ui.common.Spinner
import com.alexivanov.snapsell.ui.common.StatusChip
import com.alexivanov.snapsell.ui.theme.SnapType
import com.alexivanov.snapsell.ui.theme.snapColors

@Composable
fun HandOffScreen(
    container: AppContainer,
    listingId: String,
    onFinished: (tab: Int) -> Unit,
    onBack: () -> Unit,
) {
    val vm: HandOffViewModel = viewModel(key = listingId) { HandOffViewModel(container, listingId) }
    val state by vm.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val c = snapColors

    // Coming back from Facebook is the cue to ask how it went.
    LifecycleEventEffect(Lifecycle.Event.ON_RESUME) { vm.onResumed() }

    // API 24-28 need WRITE_EXTERNAL_STORAGE to create Pictures/Resale.
    val needsLegacyPermission = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
    val permissionLauncher = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) {
        // Proceed either way: without the album the clipboard + link still work.
        vm.stageAndOpen(context)
    }
    fun launch() {
        if (needsLegacyPermission &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
        ) {
            permissionLauncher.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        } else {
            vm.stageAndOpen(context)
        }
    }

    Box(Modifier.fillMaxSize()) {
        Scaffold(
            containerColor = c.bg,
            topBar = { SnapTopBar(title = "Hand-off", onBack = onBack) },
            bottomBar = {
                if (state.listing != null) {
                    SnapBottomBar {
                        PrimaryButton(
                            "Open Facebook Marketplace",
                            onClick = { launch() },
                            enabled = !state.staging,
                            minHeight = 58.dp,
                            modifier = Modifier.weight(1f),
                            leading = { if (state.staging) Spinner(color = c.bg) else LucideIcon(R.drawable.ic_lucide_external_link, null, size = 20.dp, tint = c.bg) },
                        )
                    }
                }
            },
        ) { padding ->
            val listing = state.listing
            if (state.loading) {
                LoadingBox(Modifier.padding(padding))
                return@Scaffold
            }
            if (listing == null) {
                Column(Modifier.padding(padding).padding(24.dp)) { Text("This listing no longer exists.", style = SnapType.body, color = c.text) }
                return@Scaffold
            }
            val n = listing.items.size

            Column(Modifier.padding(padding).fillMaxSize().verticalScroll(rememberScrollState()).padding(horizontal = 20.dp)) {
                Row(Modifier.padding(top = 20.dp), verticalAlignment = Alignment.CenterVertically) {
                    StatusChip(listing.listing.status)
                }
                Text(
                    if (state.left) "Waiting for you to come back" else "Two things are ready",
                    style = SnapType.screenHero,
                    color = c.text,
                    modifier = Modifier.padding(top = 12.dp),
                )
                Text(
                    "Marketplace won't let an app post for you, so SnapSell gets everything ready and you paste it in.",
                    style = SnapType.bodyLarge,
                    color = c.text.copy(alpha = 0.75f),
                    modifier = Modifier.padding(top = 10.dp),
                )

                // Guided steps: 2dp top rule, 1dp dividers, 28dp numbered squares, accent check when done.
                Column(Modifier.padding(top = 24.dp)) {
                    Rule(thickness = 2.dp)
                    val photos = if (state.staged) state.savedCount else n
                    Step(1, "Photos saved", "$photos cutout${if (photos == 1) "" else "s"} written to your Resale album.", done = state.staged)
                    Step(2, "Text copied", "Title and description are on the clipboard.", done = state.staged)
                    Step(3, "Paste it into Marketplace", "Facebook opens next. Pick the photos, long-press to paste.", done = state.left)
                }

                ErrorText(state.error, Modifier.padding(top = 12.dp))

                Column(Modifier.fillMaxWidth().padding(top = 24.dp).border(1.dp, c.divider).padding(14.dp)) {
                    MicroLabel("On your clipboard")
                    Text(listing.listing.title, style = SnapType.body.copy(fontWeight = FontWeight.ExtraBold), color = c.text, modifier = Modifier.padding(top = 6.dp))
                    Text(
                        listing.listing.description.lineSequence().firstOrNull().orEmpty(),
                        style = SnapType.bodySmall.copy(fontSize = 12.5.sp),
                        color = c.text.copy(alpha = 0.7f),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.padding(top = 2.dp),
                    )
                }
                Spacer(Modifier.height(24.dp))
            }
        }

        if (state.askOutcome) {
            Scrim(onClick = vm::dismissOutcome)
            Column(
                Modifier
                    .align(Alignment.BottomCenter)
                    .navigationBarsPadding()
                    .padding(16.dp)
                    .shadow(12.dp, RectangleShape, ambientColor = c.shadow, spotColor = c.shadow)
                    .background(c.bg)
                    .border(2.dp, c.text)
                    .padding(20.dp),
            ) {
                Text("Did it go up?", style = SnapType.sectionHead, color = c.text)
                Text("Tell SnapSell what happened so the item lands in the right place.", style = SnapType.body, color = c.text.copy(alpha = 0.75f), modifier = Modifier.padding(top = 6.dp, bottom = 16.dp))
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    PrimaryButton("Listed", onClick = { vm.markListed(onFinished) }, minHeight = 50.dp, modifier = Modifier.fillMaxWidth())
                    SecondaryButton("Skipped", onClick = { vm.markSkipped(onFinished) }, minHeight = 50.dp, modifier = Modifier.fillMaxWidth())
                    GhostButton("Try again", onClick = vm::dismissOutcome, minHeight = 44.dp, modifier = Modifier.fillMaxWidth())
                }
            }
        }
    }
}

@Composable
private fun Step(n: Int, title: String, body: String, done: Boolean) {
    val c = snapColors
    Column {
        Row(Modifier.fillMaxWidth().padding(vertical = 16.dp), verticalAlignment = Alignment.CenterVertically) {
            Box(
                Modifier.size(28.dp).then(if (done) Modifier.background(c.accent) else Modifier.border(1.dp, c.divider)),
                contentAlignment = Alignment.Center,
            ) {
                Text("$n", style = SnapType.rowTitle.copy(fontSize = 13.sp), color = if (done) c.bg else c.text)
            }
            Column(Modifier.weight(1f).padding(horizontal = 14.dp)) {
                Text(title, style = SnapType.rowTitle, color = c.text)
                Text(body, style = SnapType.bodySmall, color = c.text.copy(alpha = 0.65f), modifier = Modifier.padding(top = 2.dp))
            }
            if (done) LucideIcon(R.drawable.ic_lucide_check, "Done", size = 20.dp, tint = c.accent)
            else Spacer(Modifier.width(20.dp))
        }
        Rule(thickness = 1.dp)
    }
}

File: ios/SnapSell/UI/Screens/HandOff/HandOffScreen.swift
import SwiftUI

struct HandOffScreen: View {
    let listingId: String
    @Environment(AppContainer.self) private var app
    @Environment(Router.self) private var router
    @Environment(\.snap) private var c
    @Environment(\.scenePhase) private var scenePhase
    @State private var staging = false
    /// Set once Facebook was opened; the next foreground asks for the outcome.
    @State private var launched = false
    @State private var askOutcome = false
    @State private var savedCount = 0
    @State private var staged = false
    @State private var left = false
    @State private var error: String?

    private var listing: ListingWithItems? { app.inventory.listing(listingId) }

    var body: some View {
        ZStack(alignment: .bottom) {
            VStack(spacing: 0) {
                SnapTopBar("Hand-off", onBack: { router.pop() })
                if let listing {
                    content(listing)
                    VStack(spacing: 0) {
                        Rule()
                        SnapButton("Open Facebook Marketplace", systemImage: "arrow.up.right.square", enabled: !staging, loading: staging) { stageAndOpen(listing) }
                            .padding(16)
                    }
                    .background(c.bg)
                } else {
                    Text("This listing no longer exists.").snap(SnapType.body).foregroundStyle(c.text).padding(24)
                    Spacer()
                }
            }
            if askOutcome {
                Scrim().onTapGesture { dismissOutcome() }
                VStack(alignment: .leading, spacing: 0) {
                    SectionHead("Did it go up?")
                    Text("Tell SnapSell what happened so the item lands in the right place.").snap(SnapType.body).foregroundStyle(c.text.opacity(0.75)).padding(.top, 6).padding(.bottom, 16)
                    VStack(spacing: 8) {
                        SnapButton("Listed") { setStatus(.listed, tab: 1) }
                        SnapButton("Skipped", kind: .secondary) { setStatus(.skipped, tab: 0) }
                        SnapButton("Try again", kind: .ghost) { dismissOutcome() }
                    }
                }
                .padding(20).background(c.bg).overlay(Rectangle().stroke(c.text, lineWidth: SnapMetrics.rule))
                .shadow(color: .black.opacity(0.25), radius: 12, y: 6)
                .padding(16)
            }
        }
        .onChange(of: scenePhase) { _, phase in
            // Coming back from Facebook is the cue to ask how it went.
            if phase == .active, launched { askOutcome = true }
        }
        .snapScreen()
    }

    private func content(_ listing: ListingWithItems) -> some View {
        let n = listing.items.count
        let photos = staged ? savedCount : n
        return ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                StatusChip(status: listing.listing.status).padding(.top, 20)
                Text(left ? "Waiting for you to come back" : "Two things are ready").snap(SnapType.screenHero).foregroundStyle(c.text).padding(.top, 8)
                Text("Marketplace won't let an app post for you, so SnapSell gets everything ready and you paste it in.")
                    .snap(SnapType.bodyLarge).foregroundStyle(c.text.opacity(0.75)).padding(.top, 6)
                VStack(spacing: 0) {
                    Rule()
                    step(1, "Photos saved", "\(photos) cutout\(photos == 1 ? "" : "s") written to your Resale album.", done: staged)
                    step(2, "Text copied", "Title and description are on the clipboard.", done: staged)
                    step(3, "Paste it into Marketplace", "Facebook opens next. Pick the photos, long-press to paste.", done: left)
                }
                .padding(.top, 24)
                if let error { ErrorText(error).padding(.top, 12) }
                VStack(alignment: .leading, spacing: 0) {
                    MicroLabel("On your clipboard")
                    Text(listing.listing.title).font(.custom("Archivo-ExtraBold", size: 14)).foregroundStyle(c.text).padding(.top, 6)
                    Text(listing.listing.description.split(separator: "\n", omittingEmptySubsequences: false).first.map(String.init) ?? "")
                        .font(.custom("Archivo-Regular", size: 12.5)).foregroundStyle(c.text.opacity(0.7)).lineLimit(1)
                }
                .padding(14).frame(maxWidth: .infinity, alignment: .leading)
                .overlay(Rectangle().stroke(c.divider, lineWidth: 1))
                .padding(.top, 24)
                Spacer(minLength: 24)
            }
            .padding(.horizontal, 20)
        }
    }

    private func step(_ n: Int, _ title: String, _ body: String, done: Bool) -> some View {
        VStack(spacing: 0) {
            HStack(spacing: 14) {
                Text("\(n)").font(.custom("Archivo-SemiBold", size: 13)).foregroundStyle(done ? c.bg : c.text)
                    .frame(width: 28, height: 28)
                    .background(done ? c.accent : .clear)
                    .overlay(Rectangle().stroke(done ? .clear : c.divider, lineWidth: 1))
                VStack(alignment: .leading, spacing: 2) {
                    Text(title).snap(SnapType.rowTitle).foregroundStyle(c.text)
                    Text(body).snap(SnapType.bodySmall).foregroundStyle(c.text.opacity(0.65))
                }
                Spacer(minLength: 0)
                if done { Image(systemName: "checkmark").font(.system(size: 18, weight: .bold)).foregroundStyle(c.accent) } else { Spacer().frame(width: 20) }
            }
            .padding(.vertical, 16)
            Rule(color: c.divider).frame(height: 1)
        }
    }

    /// Album + clipboard + open Facebook.
    private func stageAndOpen(_ listing: ListingWithItems) {
        guard !staging else { return }
        staging = true
        error = nil
        Task {
            var files: [String] = []
            for item in listing.items {
                files.append(item.photoFile)
                // The original photo shows the item in context; buyers like both.
                if item.originalPhotoFile != item.photoFile { files.append(item.originalPhotoFile) }
            }
            let store = app.photoStore
            let images = await Task.detached { files.compactMap { store.load($0) } }.value
            let outcome = await ResaleAlbum.save(images)
            let saved = outcome == .denied ? 0 : images.count
            ClipboardStager.stage(title: listing.listing.title, price: listing.listing.price, description: listing.listing.description)
            let route = await MarketplaceLauncher.open()
            staging = false
            launched = route != .none
            staged = true
            left = route != .none
            savedCount = saved
            error = route == .none ? "No app could open Facebook Marketplace." : (outcome == .denied ? "Photos weren't saved: allow photo access in Settings." : nil)
        }
    }

    private func dismissOutcome() { askOutcome = false; launched = false }

    /// Listed lands on the Listings tab (1); Skipped on Items (0).
    private func setStatus(_ status: ListingStatus, tab: Int) {
        try? app.inventory.updateListingStatus(listingId, status: status)
        askOutcome = false
        launched = false
        router.goHome(tab: tab)
    }
}
